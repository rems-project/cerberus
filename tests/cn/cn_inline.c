

enum size {
  big,
  medium,
  small,
};

/*@ function (i32) lookup_size_shift_cn (u32 sz) @*/

static inline int
lookup_size_shift (enum size sz)
/*@ cn_function lookup_size_shift_cn;
    ensures return == lookup_size_shift_cn(sz); @*/
{
  switch (sz) {
    case big:
      return 12;
    case medium:
      return 8;
    case small:
      return 2;
    default:
      /* shouldn't happen */
      return 0;
  }
}

/* The original plan was to inline the above into the below, but when possible
   it is much simpler to use its functional representation, as done here. */

int
f (void)
/*@ ensures return < (1000i32); @*/
{
  int x;
  x = 3 * lookup_size_shift(medium);
  x += 5 * lookup_size_shift(small);
  return x;
}

int main(void)
/*@ trusted; @*/
{
  int r = f();
}
