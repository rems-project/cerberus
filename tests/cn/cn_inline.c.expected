return code: 0
tests/cn/cn_inline.c:13:5: warning: experimental keyword 'cn_function' (use of experimental features is discouraged)
/*@ cn_function lookup_size_shift_cn; @*/
    ^~~~~~~~~~~ 
[1/2]: lookup_size_shift -- pass
[2/2]: f -- pass
